src/connector.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const net = require('net');
+const lookupAll = require('dns-lookup-all');
+
+class Connector {
+  constructor(options, multiSubnetFailover) {
+    this.options = options;
+    this.multiSubnetFailover = multiSubnetFailover;
+  }
+
+  execute(cb) {
+    if (net.isIP(this.options.host)) {
+      this.executeForIP(cb);
+    } else {
+      this.executeForHostname(cb);
+    }
+  }
+
+  executeForIP(cb) {
+    const socket = net.connect(this.options);
+
+    const onError = (err) => {
+      socket.removeListener('error', onError);
+      socket.removeListener('connect', onConnect);
+
+      socket.destroy();
+
+      cb(err);
+    };
+
+    const onConnect = () => {
+      socket.removeListener('error', onError);
+      socket.removeListener('connect', onConnect);
+
+      cb(null, socket);
+    };
+
+    socket.on('error', onError);
+    socket.on('connect', onConnect);
+  }
+
+  executeForHostname(cb) {
+    lookupAll(this.options.host, 4, (err, addresses) => {
```

Comment:
Ah, you're right - there's no specific reason why we only support IPv4.

What we should have is a global option on whether you want to use IPv4 *or* IPv6. Otherwise, if your servers support both, we'll try connecting to each one of them twice.